### DISCUSSION TITLE
Data Modification with Multiple Labels

### DISCUSSION THREAD

**Discussion by joshuatbland** [LaboratoireMecaniqueLille/crappy#99] [Q&A] [answered]
Hello everyone! I am new to crappy and I am trying to learn more about on-the-fly data modification capabilities. It looks like modifiers are the best way to do real time calculations with single data labels. What would be the best way to do on-the-fly calculations with multiple labels (like taking the average or mean of several labels at every time point)? Any insight into this would be greatly appreciated. Thanks.

**Comment by WeisLeDocto** [comment, marked as answer]
Hi !

It all depends if the labels come from a same Block, or from different ones.

* If they come from a same Block, it is possible to perform any operation on all the labels at once. Basic example:
   ```python
   class MyModifier(crappy.modifier.Modifier):
       def __call__(self, data):
           data['avg'] = 0.5 * (data['label_1'] + data['label_2'])
           return data
   ```
* If they come from different Blocks, things get much trickier and there is no universal answer. What's for sure is that a `Modifier` alone cannot do the job, a `Block` needs to be used. In the general case, the labels will not be generated in a synchronized way by the several Blocks, so it's impossible to define operations to compute *at every time point*. One can define time-based operations (e.g. average all input labels together over a given period of time), but they won't run *at every time point*.

   That being said, Crappy comes with two Blocks that can synchronize signals from several input Blocks into a single output on a common time base. This output can then be processed by a regular `Modifier`, as described in my first point. The first Block is the [`MeanBlock`](https://crappy.readthedocs.io/en/v2.0.4/crappy_docs/blocks.html#mean-block), that averages the received labels over a given delay (1 average per label), and returns all the averages. It's recommended when a low data rate is acceptable (<10 SPS). The second Block is the [`Multiplexer`](https://crappy.readthedocs.io/en/v2.0.4/crappy_docs/blocks.html#multiplexer), that performs linear interpolation on the received labels to output data at a target frequency. It's indicated when a higher data rate is needed. 

   Note that with both options the original data is lost when averaging or interpolating. You'll find examples for both Blocks in the [`examples` folder](https://github.com/LaboratoireMecaniqueLille/crappy/tree/f7ceb73e95cc42737005649b843c01a9859873d3/examples/blocks) on GitHub.

**Reply by joshuatbland** [reply]
Ok, great - thank you for the response. This was very helpful!
